File: output_src/xml_model.py

from lxml.etree import XML, tostring, parse, Element, XMLParser
from fd.util.xml import xml_element_from_file, xml_element_to_string


class XmlModel( object ):
    """
    XmlModel is a base class for all concrete models (and their containers) that have XML representation.

    XmlModels are entirely, internally represented with a single, in-memory lxml element. So, all accessors manipulate
    and read from the single lxml element. With this design, you can still manipulate many objects at once with the
    traditional lxml iterators, but still relate to the XML element as an object.

    All models "must" expect the self.element member variable to be accessed publicly.

    These models are intended to be in-memory and are only, incidentally, able to move to or from xml files, string or
    other lxml elements. These models are explicitly not part of an ORM (that is, they're not designed primarily to be
    mapped to a particular database; they're designed to represent XML in memory).
    """


    def __init__( self, xml_element = None, filename = None, xml_string = None ):
        """
        Constructor.
        May be constructed by:

            1. Reading from a file.
                OR
            2. Parsing an xml string.
                OR
            3. Enveloping an existing lxml element.
                OR
            4. By constructing the default empty element in the derived class' constructor. See xml.model.user.User.
               This is detected in the derived class constructor by self.element equalling None after this constructor
               has been invoked.

        in that order.
        """

        if filename is not None:

            self.element = xml_element_from_file( filename )

        elif xml_string is not None:

            self.element = XML( xml_string, XMLParser(remove_blank_text=True) )

        elif xml_element is not None:

            self.element = xml_element

        else:

            self.element = None



    def as_string( self ):
        """
        Serializes this xml element and returns it as a string.
        @return string
        """

        return xml_element_to_string( self.element )



    def save_to_file( self, full_filepath ):
        """
        Serializes this xml element and saves it to the provided filename (must be a full pathname).
        """

        with open( full_filepath, 'w+' ) as file:
            file.write( self.as_string() )



    def get_tag_by_name( self, tag_name, parent_element = None ):
        """
        Retrieves a single lxml element, that is a subelement of the main self.element, by tag name. If the tag does
        not exist, it will return None. Please use XmlModel.get_or_create_tag_by_name if you want it to automatically
        create it. If parent_element is not None, it will use that as the reference parent instead of self.element.
        """

        if parent_element is None:
            parent_element = self.element

        tag_name = tag_name.encode('utf-8')
        tag_element = parent_element.find( tag_name )

        if tag_element is None:
            return None

        return tag_element



    def get_or_create_tag_by_name( self, tag_name, parent_element = None ):
        """
        Retrieves a single lxml element, that is a subelement of the main self.element, by tag name. If the tag does
        not exist, it will create it and return it. If parent_element is not None, it will use that as the reference
        parent instead of self.element.
        """

        if parent_element is None:
            parent_element = self.element

        tag_name = tag_name.encode('utf-8')
        tag_element = parent_element.find( tag_name )

        if tag_element is None:
            tag_element = Element( tag_name )
            parent_element.append( tag_element )

        return tag_element



    def set_attribute_value( self, attribute_name, attribute_value ):
        """
        Sets an xml attribute on self.element.
        """

        self.element.attrib[ attribute_name.encode('utf-8') ] = attribute_value.encode( 'utf-8')



    def get_attribute_value( self, attribute_name ):
        """
        Retrieves the value of the given attribute on self.element.
        Returns None if the attribute does not exist.
        """

        try:

            return self.element.attrib[ attribute_name.encode('utf-8') ]

        except Exception, e:

            return None



    def set_tag_value( self, tag_name, tag_value, parent_element = None ):
        """
        Sets the xml text() of a given subelement of the main self.element, by tag name.
        If the tag does not exist, it will create it and set the text for that subelement.
        If parent_element is not None, it will use that as the reference parent instead of self.element.
        """

        if parent_element is None:
            parent_element = self.element

        tag_element = self.get_or_create_tag_by_name( tag_name, parent_element = parent_element )
        tag_element.text = tag_value.encode('utf-8')



    def get_tag_value( self, tag_name, parent_element = None ):
        """
        Retrieves the xml text() of a given subelement of the main self.element, by tag name, as a string.
        If the tag does not exist, it will return None.
        If parent_element is not None, it will use that as the reference parent instead of self.element.
        """

        if parent_element is None:
            parent_element = self.element

        tag_name = tag_name.encode('utf-8')
        tag_element = parent_element.find( tag_name )

        if tag_element is None:
            return None

        return tag_element.text

File: output_src/XmlModel.js
XmlModel = function( data_reference ){

    var me = this;

    var init = function(){
    };


    /**
     * Returns a (reference) object to the attribute with a name matching attribute_name
     *
     * @param string attribute_name
     * @returns Object | null
     */
    me.getAttribute = function( attribute_name ){

        if( data_reference._attributes ){
            for( i in data_reference._attributes ){

                var attribute = data_reference._attributes[i];

                if( attribute._name && attribute._name == attribute_name ) {
                    return attribute;
                }

            }
        }

        return null;

    };


    /**
     * Returns a the value of an attribute with a name matching attribute_name
     *
     * @param string attribute_name
     * @returns string | null
     */
    me.getAttributeValue = function( attribute_name ){

        var attribute = me.getAttribute( attribute_name );

        if( attribute && attribute._value ){
            return attribute._value;
        }

        return false;

    };


    /**
     * Sets the value of an attribute. If the attribute doesn't exist, it will be created.
     *
     * @param string attribute_name
     * @param string attribute_value
     * @returns self
     */
    me.setAttributeValue = function( attribute_name, attribute_value ){

        var attribute = me.getAttribute( attribute_name );

        // create attribute if it doesn't exist
        if( !attribute ){

            data_reference._attributes.push({
                _name: attribute_name,
                _value: attribute_value
            });

        } else {

            attribute._value = attribute_value;

        }

        return me;

    };


    /**
     * Removes an attribute by name.
     *
     * @param string attribute_name
     * @returns self
     */
    me.removeAttribute = function( attribute_name ){

        if( data_reference._attributes ){
            for( i in data_reference._attributes ){

                var attribute = data_reference._attributes[i];

                if( attribute._name && attribute._name == attribute_name ) {

                    data_reference._attributes.splice( i, 1);

                }

            }
        }

        return me;

    };


    /**
     * Returns a (reference) object to a child by tag.
     * If more than one child with the same tag exists, the first one is returned.
     *
     * @param string tag
     * @returns Object | null
     */
    me.getChild = function( tag ){

        if( data_reference._children ){
            for( i in data_reference._children ){

                var child = data_reference._children[i];

                if( child._tag && child._tag == tag ) {
                    return child;
                }

            }
        }

        return null;

    };


    /**
     * Returns the value of a text child or null if it doesn't exist or if it is not a text child.
     *
     * @param string tag
     * @returns string | null
     */
    me.getChildTextValue = function( tag ){

        var child = me.getChild( tag );

        if( child && child._text ){
            return child._text;
        }

        return null;

    };


    /**
     * Sets the value of a text child. If it doesn't exist, it will be created.
     *
     * @param string child_tag
     * @param string child_text_value
     * @returns self
     */
    me.setChildTextValue = function( child_tag, child_text_value ){

        var child = me.getChild( child_tag );

        // create child if it doesn't exist
        if( !child ){

            data_reference._children.push({
                _tag: child_tag,
                _attributes: [],
                _text: child_text_value
            });

        } else {

            child._text = child_text_value;

        }

        return me;

    };


    /**
     * Removes a child by tag.
     *
     * @param string tag
     * @returns self
     */
    me.removeChild = function( tag ){

        if( data_reference._children ){
            for( i in data_reference._children ){

                var child = data_reference._children[i];

                if( child._tag && child._tag == tag ) {

                    data_reference._children.splice( i, 1);

                }

            }
        }

        return me;

    };


    /**
     * Get children by tag
     *
     * @param string tag
     * @returns self
     */
    me.getChildrenByTag = function( tag ){

        var children = [];

        if( data_reference._children ){
            for( i in data_reference._children ){

                var child = data_reference._children[i];

                if( child._tag && child._tag == tag ) {

                    children.push( data_reference._children[i] );

                }

            }
        }

        return children;

    };


    /**
     * Get the text value of this object.
     *
     * @returns string
     */
    me.getTextValue = function(){

        if( data_reference._text && typeof(data_reference._text) ){
            return data_reference._text;
        }

        return null;

    };


    /**
     * Set the text value of this object.
     *
     * @param string text_value
     * @returns string
     */
    me.setTextValue = function( text_value ){

        data_reference._text = text_value;

        return me;

    };

    /**
     * Remove the text value of this object.
     *
     * @returns string
     */
    me.removeTextValue = function(){

        data_reference._text = null;

        return me;

    };


    init();

};